cpp/include/cudf/utilities/span.hpp

```diff
@@ -278,7 +278,10 @@ struct host_span : public cudf::detail::span_base<T, Extent, host_span<T, Extent
    * @param idx the index of the element to access
    * @return A reference to the idx-th element of the sequence, i.e., `data()[idx]`
    */
-  constexpr typename base::reference operator[](size_type idx) const { return this->_data[idx]; }
+  CUDF_HOST_DEVICE constexpr typename base::reference operator[](size_type idx) const
```

Comment:
Let me double-check why this is necessary. Requiring a host span to be accessible from the device doesn't seem correct.

Reply:
There are places where we need to access a device span from the host or a host span from the device, e.g. https://github.com/rapidsai/cudf/blob/bbf4f7824c23c0c482f52bafdf1ece1213da2f65/cpp/src/io/parquet/writer_impl.cu#L1971. 

Question to reviewers. Do you know if this is a legit use?

Reply:
This is a bit iffy. IMO we should rewrite this as something like
`row_group_fragments.device_view().flat_view() + flattened_index` to avoid using the index operator on the host. We might want to add an API to `device_2dspan` to get an address of an element at [x][y]. Unlike the actual index operator, this is valid to do on the host.

Reply:
I’ve opened https://github.com/rapidsai/cudf/issues/17797 to track the `constexpr` issues with spans. I’ll revert all span-related changes in this PR.